qiskit/ignis/experiments/calibration/instruction_data/compiler.py

```diff
@@ -0,0 +1,318 @@
+        """Create new parser.
+
+        Args:
+              inst_def: Instruction definition object.
+        """
+        self._inst_def = inst_def
+
+        # these parameters are set on the fly
+        # TODO remove them
+        self._gate_id = None
+        self._free_parameters = []
+
+    def __call__(self,
+                 source: str,
+                 gate_id: str,
+                 free_parameters: Optional[List[str]] = None) -> pulse.Schedule:
+        """Parse source code and return Schedule.
+
+        TODO remove gate_id and free parameters from parser.
+        Blocker is parametrization of pulse duration.
+        If we can parametrize the duration, we can populate schedule with unbound parameters and
+        parser doesn't need to call parameter table to get integer duration.
+        Thus everything can be correctly offloaded the parameter binding operation
+        to the get_gate_schedule method that calls this parser.
+
+        Args:
+            source: Calibration DSL representing a pulse schedule.
+
+        Returns:
+            Pulse schedule object.
+        """
+        self._gate_id = gate_id
+        self._free_parameters = free_parameters or []
+
+        return self.visit(parse(source))
+
+    def _get_channel(self, ch_str: str) -> pulse.channels.Channel:
+        """A helper function to convert channel string into Qiskit object.
+
+        Args:
+            ch_str: String representation of channel.
+
+        Returns:
+            Qiskit pulse channel object.
+        """
+        match = self.chan_regex.match(ch_str)
+        if match:
+            prefix, index = match.group(1), int(match.group(2))
+            channel = self._inst_def.pulse_channels[prefix].value
+            return channel(index=index)
+
+        raise Exception('Channel name {name} is not correct syntax.'.format(name=ch_str))
+
+    def visit(self, node):
+        """Visit a node."""
+        method = 'visit_' + node.__class__.__name__
+        visitor = getattr(self, method)
+
+        return visitor(node)
+
+    def visit_PulseInst(self, node: PulseInst) -> pulse.Instruction:
+        """Evaluate pulse instruction node and return play instruction.
+
+        Args:
+            node: PulseInst node to evaluate.
+
+        Returns:
+            Play instruction.
+        """
+        parametric_pulse = self._inst_def.parametric_shapes[node.shape].value
+
+        # generate parameter names on the fly
+        parametric_pulse_kwargs = dict()
+        for pname in utils.get_pulse_parameters(parametric_pulse):
+            scoped_pname = utils.add_scope(pname, node.channel, node.name)
+            # TODO calling parameter biding here is bit strange.
+            # This parameter binding should be offloaded to the get_gate_schedule method.
+            # However, we cannot do this now because of parametrization of duration.
+            # The role of parser is to create program, not the parameter-bound schedule.
```

Comment:
I agree with you.
Can we, for now, only work with non-parametric durations?

Reply:
yes. strictly speaking this value should be cast into integer.